server/src/main/java/org/elasticsearch/search/fetch/subphase/FieldFetcher.java

```diff
@@ -80,93 +83,71 @@ static FieldFetcher create(SearchExecutionContext context,
-    }
 
-    private static boolean isNestedObjectSubfield(String field, Set<String> nestedFields) {
-        for (String nestedField : nestedFields) {
-            if (field.startsWith(nestedField)) {
-                return true;
-            }
+            // add a special ValueFetcher that filters source and collects its subfields
+            fieldContexts.put(
+                nestedFieldPath,
+                new FieldContext(nestedFieldPath, new NestedValueFetcher(nestedFieldPath, nestedSubFieldFetcher))
+            );
         }
-        return false;
-    }
 
-    /**
-     * Given a list of requested field patterns, this method determines the set of
-     * set of available field mappers that are needed to handle fetching the subfields.
-     *
-     * Given two available nested field mappings "foo" and "foo.bar", we want to return the
-     * one closest to the root document (in this case "foo") for any subfield of foo.*, so patterns like
-     * "*", "foo.*", "foo.bar.baz" should all just return the nested field "foo" that further recursive sub-processing is routed too.
-     */
-    static Set<String> determineMatchingNestedFieldPaths(
-        Collection<FieldAndFormat> fieldAndFormats,
-        List<String> nestedMappingPaths
-    ) {
-        Set<String> matchingNestedFields = new HashSet<>();
-        for (FieldAndFormat fieldAndFormat : fieldAndFormats) {
-            String fieldPattern = fieldAndFormat.field;
-            // sort available nested mappings paths according to their path length, so we process shortest prefixes first
-            nestedMappingPaths.sort(Comparator.comparing(String::length));
-            for (String nestedField : nestedMappingPaths) {
-                if (Regex.simpleMatch(fieldPattern, nestedField) || fieldPattern.startsWith(nestedField)) {
-                    // have we already added another nested field path that is a prefix of this?
-                    if (matchingNestedFields.stream().anyMatch(s -> nestedField.startsWith(s)) == false) {
-                        matchingNestedFields.add(nestedField);
-                    }
-                }
-            }
+        CharacterRunAutomaton unmappedFieldsFetchAutomaton = null;
+        if (unmappedFetchPattern.isEmpty() == false) {
+            unmappedFieldsFetchAutomaton = new CharacterRunAutomaton(
+                Regex.simpleMatchToAutomaton(unmappedFetchPattern.toArray(new String[unmappedFetchPattern.size()]))
+            );
         }
-        return matchingNestedFields;
+        return new FieldFetcher(fieldContexts, unmappedFieldsFetchAutomaton, fieldsInsideNested.keySet());
     }
 
     private final Map<String, FieldContext> fieldContexts;
-    private final CharacterRunAutomaton unmappedFetchAutomaton;
-    private final boolean includeUnmapped;
+    private final CharacterRunAutomaton unmappedFieldsFetchAutomaton;
+    private final Set<String> excludedNestedPaths;
```

Comment:
I wonder if we could simplify by removing this extra variable and instead check `fieldContexts#containsKey` here since it contains the nested paths? This may also fix the weird behavior with flattened where `include_unmapped` returns the subkeys of the flattened object.

Reply:
Interesting. I was going to say that fieldContexts contain more than just the nested field paths, but then again all keys _should_ either be mapped leaf values (that we don't want to traverse when getting unmapped fields) or nested paths. So I'll give this a try and see if it breaks any existing tests.

Reply:
Seems to work nicely so far